src/cascadia/WindowsTerminal/IslandWindow.cpp

```diff
@@ -658,6 +658,56 @@ long IslandWindow::_calculateTotalSize(const bool isWidth, const long clientSize
         }
         break;
     }
+
+    case WM_QUERYENDSESSION:
+    {
+        // For WM_QUERYENDSESSION and WM_ENDSESSION, refer to:
+        //
+        // https://docs.microsoft.com/en-us/windows/win32/rstmgr/guidelines-for-applications
+        //
+        // The OS will send us a WM_QUERYENDSESSION when it's preparing an
+        // update for our app. It will then send us a WM_ENDSESSION, which gives
+        // us a small timeout (~30s) to actually shut down gracefully. After
+        // that timeout, it will send us a WM_CLOSE. If we still don't close
+        // after the WM_CLOSE, it'll force-kill us (causing a crash which will be
+        // bucketed to MoAppHang).
+
+        // If we need to do anything to prepare for being told to shutdown,
+        // start it here.
+        //
+        // If (in the future) we need to prevent logoff, we can return false here.
+
+        TraceLoggingWrite(
+            g_hWindowsTerminalProvider,
+            "QueryEndSession",
+            TraceLoggingDescription("Emitted when the OS has begun the process of updating the Terminal"),
+            TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+            TraceLoggingKeyword(TIL_KEYWORD_TRACE));

```

Comment:
Isn't this event redundant with the one for `WM_ENDSESSION`? If we return `true` here, then almost always we should also receive a `WM_ENDSESSION` event anyways right?
In that case you could remove this entire `case` because the `DefWindowProc` already returns `true`.

Reply:
Meh, yea probably. I was thinking originally that we would shove some logic in there, but turns out we really don't need it.